docs/tech-notes/index-backfill.md

```diff
@@ -0,0 +1,156 @@
+
+In this case, the index entry found by `InitPut` has never been written to.
+Since we know that all operations on the table at or after `tScan` cause updates
+to the index, an absent index entry indicates that the primary table entry has
+not been updated since `tScan` and is therefore safe to echo to the index.
+
+For example, when the backfiller encounters the pair `1` `a`, the corresponding
+index entry is empty, so it proceeds to write.
+
+| `k` | `v` | `t` (MVCC timestamp) | `idx_k` (`v`) | `idx_v` (`k`) | `idx_t` (idx MVCC timestamp) |
+|-----|-----|----------------------|---------------|---------------|------------------------------|
+| `1` | `a` | 0                    | `a`           | `1`           | `after tScan`                |
+| `2` | `b` | `tWrite+1`           | `b`           | `2`           | `tWrite+1`                   |
+| `3` | `c` | 0                    | `c`           | `(tombstone)` | `tScan+1`                    |
+| `3` | `d` | `tScan+1`            | `d`           | `3`           | `tScan+1`                    |
+| `4` | `e` | 0                    |               |               |                              |
+| `5` | `e` | `tScan+1`            | `e`           | `5`           | `tScan+1`                    |
+| `6` | `f` | 0                    |               |               |                              |
+| `6` | ` ` | `tScan+1`            | `f`           | `(tombstone)` | `tScan+1`                    |
+| `7` | `g` | `0`                  |               |               |                              |
+| `8` | `g` | `tScan+1`            | `g`           | `8`           | `tScan+1`                    |
+
+### The `InitPut` sees a non-empty entry with identical value bytes
+
+In this case, the index entry found by `InitPut` has a value identical to the
+key of the table. There are just two ways this could happen, both of which are
+ok:
+
+1. A table entry was written after `tWrite` but before `tScan`, leaving behind
+   a corresponding index entry. This case is safe since we're just repeating
+   previous work.
+2. A backfiller dies before it checkpoints, leaving behind an index entry. When
+   the backfiller restarts, it encounters its previously written entry. This
+   case is safe since we're again just repeating previous work.
+
+There's never a case where this would accidentally cause a uniqueness
+constraint violation, since the index entry contains the (guaranteed unique)
+primary key of the table in the value. A uniqueness constraint violation always
+manifests as an unexpected value for a key.
+
+For example, when the backfiller encounters the pair `2` `b`, the corresponding
+index entry is `b` `2`, so the backfiller does nothing and moves on.
+
+## The `InitPut` fails
+
+In all other cases, the `InitPut` will fail. In these cases, we will retry the
+operation transactionally with a fresh read to determine whether to return a
+uniqueness constraint violation or to write a more up to date value.
```

Comment:
The InitPut *must* always write a value at `tScan`. As an optimization you could consider checking if there is a value written at timestamp <= `tScan` that differs from what you are writing (tombstones can be ignored), so that you can detect if there was a UNIQUE constraint violation at timestamp `tScan`, but that doesn't save you from needing to do the verification at the end because in the interim there could have been a constraint violation. Well, not quite, you could also check if the value lies between two tombstones (in time), but I'm not sure how hard that will be. That will allow you to not need to do the verification at the end. Precisely, an InitPut is not allowed to initialize a key at `tScan` if it is overriding another value, or has already been overridden by another value (where values are not tombstones). This takes care of the two cases 1. constraint violation at `tScan` and 2. constraint violation at T > `tScan` which was not seen because the backfill value at `tScan` was not written when the transaction at time T wrote the index.

The timestamp cache can be ignored, because the index is not readable at the moment.